Add ArrayView util for sorting, filtering, and paginating Arrays

blueridger opened the pull request:
# Notes
This is a utility Array wrapper that I am creating for the Explorer Home react view, but I expect this will continue to be useful for any page that has a filtered+paginated+sorted table. The purpose is to wrap the state management needed to apply those 3 transformations, in order to simplify both the UI code and the data construction code.

The pagination and filtering here is solely for UI presentation. It does not offer memory savings since the entire array is still stored in memory (and additional state data is stored on top).

# Automated Tests
`yarn test`
added thorough unit tests

# Manual Testing
None

decentralion commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+    order: "ascending" or "descending"
+   */
+  sortBy(pluckArg: PluckFn<T>, order: "ascending" | "descending") {
+    this._array = sortBy(this._array, pluckArg);
+    this._filteredArray = sortBy(this._filteredArray, pluckArg);
+    if (order === "descending") {
+      this._array.reverse();
+      this._filteredArray.reverse();
```
Hmm, sort is `O(n * log n)`, and this function is likely to be in the hot path for re-rendering the frontend (i.e. user changes the sort order in the UI). This function does two sorts, which is unnecessary. In particular, we never need to sort the full array since we don't actually return directly from it, we just need to sort the filtered array. It's easy to imagine cases where the filtered array is dramatically smaller than the full array (e.g. filtered down to just contributions in October 2020, or filtered down to just GitHub pull requests), so this could be a real change in performance.

How about we cache the sortBy arguments, so that way when the filter changes, we can compute a new filteredArray and then re sort it?

blueridger replied:
Good idea, I'll add a cache.

decentralion reviewed (commented):
I'd like to see an example of migrating existing code to use this interface before using it. I wonder about how cleanly it will integrate... as an example, the rowsPerPage is stored inside this class, but is not accessible. However, the UI using this class will want to know the rowsPerPage setting so that it can render it and update appropriately. In that case, will the containing UI also have its own representation of the rowsPerPage state? Are there potential bugs where these two pieces of state desynchronize? Could we avoid those issues by having this class expose rowsPerPage? etc

decentralion commented on src/util/sortBy.js:
```diff
@@ -1,6 +1,6 @@
 //@flow
 
-type PluckFn<T> = (T) => any;
+export type PluckFn<T> = (T) => any;
```
Probably shouldn't be a part of this PR, but I wonder if we should change the type signature to `(T) => number | string`. I can't really think of a case where trying to compare on something other than numbers or strings is a good idea.

blueridger replied:
I was wondering that too

blueridger commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+  _array: Array<T>;
+  _filteredArray: Array<T>;
+  _rowsPerPage: number;
+  _pages: number;
```
sure

blueridger commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+  setRowsPerPage(rowsPerPage: number): number {
+    rowsPerPage = Math.min(rowsPerPage, this._filteredArray.length);
+    rowsPerPage = Math.max(rowsPerPage, 1);
+    rowsPerPage = parseInt(rowsPerPage.toFixed(0));
```
This line is to remove decimals since JS doesn't have an int type, because lol I forgot about Math.round
decimals would cause not good things to happen later on if we stored them, so this is just for robustness.

decentralion replied:
`parseInt(1.2)` already returns `1` tho

blueridger replied:
today I learned

blueridger commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+  _pages: number;
+
+  constructor(data: $ReadOnlyArray<T>) {
+    this._array = JSON.parse(JSON.stringify(data)); //Deep Copy
```
accepted

blueridger commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+    return this._pages;
+  }
+
+  /**
+    Get the contents of the specified page.
+   */
+  getPage(pageNumber: number): $ReadOnlyArray<T> {
+    pageNumber = parseInt(pageNumber.toFixed(0));
+    if (pageNumber < 1 || pageNumber > this._pages) return [];
```
sure, will change to an error.

decentralion replied:
Are you keeping the 1-based indexing? if so, why?

blueridger replied:
Because the page number is ultimately a UI concept, which will be 1 indexed on the page selector front end element, it seems intuitive to me for `page` to be 1-indexed whereas `index` would be 0-indexed.

decentralion commented on src/util/arrayView.js:
```diff
@@ -0,0 +1,74 @@
+import {type PluckFn} from "./sortBy";
+import sortBy from "./sortBy";
+
+export class ArrayView<T> {
```
Another thing: Let's have a module docstring explaining what the intention is for this module. The context that it's specifically intended for displaying data in frontends would be helpful.

Since it's intended for frontend use only, we should probably put it in `src/webutil` instead of `src/util`

blueridger replied:
sure, will move it and add a doc comment.

blueridger commented:
> I'd like to see an example of migrating existing code to use this interface before using it. I wonder about how cleanly it will integrate... as an example, the rowsPerPage is stored inside this class, but is not accessible. However, the UI using this class will want to know the rowsPerPage setting so that it can render it and update appropriately. In that case, will the containing UI also have its own representation of the rowsPerPage state? Are there potential bugs where these two pieces of state desynchronize? Could we avoid those issues by having this class expose rowsPerPage? etc

So d3 won't need to track rows per page for the table building, though it will track it separately for the selection mechanism. The data flow on this attribute is generally unidirectional from the front end to the mdule. The only reason the front end might need it exposed by this module is for the initial state, which practically speaking is just "All pages" rather that the actual number, so I'm skeptical how useful exposing it will be. The cool thing about this module, is that it is easily tweakable and upgradable if we change our minds during implementation.

I will look around for somewhere we could integrate it into existing code.

blueridger commented:
Addressed feedback, and added a setData method to be able to change the data set, in the case that external filtering happens like in the ExplorerHomeDataBuilder.

decentralion commented:
> So d3 won't need to track rows per page for the table building, though it will track it separately for the selection mechanism.

I'm thinking about React moreso than d3.

An example of an existing UI is a table with a rowsPerPage selection is the ledger history viewer: https://cred.sourcecred.io/#/ledger

An example of a UI that would really benefit from being made into a table with sorting, searching, and pagination is the Grain Account viewer: https://cred.sourcecred.io/#/accounts

Either could be a good test case for demonstrating the suitability of this util.

BrianBelhumeur commented on src/webutil/arrayState.js:
```diff
@@ -0,0 +1,104 @@
+  filter(filterFn?: (T) => boolean): number {
+    this._filterFn = filterFn;
+    this._filteredArray = filterFn ? this._array.filter(filterFn) : this._array;
+    if (this._sortArgs) this.sortBy(...this._sortArgs);
```
Have you run prettier on this file?

decentralion replied:
CI wouldn't be passing if the code weren't prettified

BrianBelhumeur commented on src/webutil/arrayState.test.js:
```diff
@@ -0,0 +1,231 @@
+        expect(arrayState.getPage(2)).toEqual([{num: 1, str: "1"}]);
+      });
+
+      it("should create no less that 1 page", () => {
```
```suggestion
      it("should create no less than 1 page", () => {
```

blueridger commented:
So I dived into React, and I'm seeing how this doesn't integrate super well. Going to do some more learning on React, and maybe come back to something like this with that knowledge.
